File: benchmarks/livermore/k08_adi.c
/* K8 — ADI Integration (1-based indexing, matches Fortran exactly)
   Fortran: DIMENSION U1(2,101,5), U2(2,101,5), U3(2,101,5)
            DIMENSION DU1(101), DU2(101), DU3(101)
   N=100, NL1=1, NL2=2, KX=2..3, KY=2..N-1
   Fortran column-major: U1(K1,K2,K3) at (K3-1)*2*101 + (K2-1)*2 + K1
   (1-based flat index) */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define NDIM  101
#define N     100
#define KXD   5
#define NREPS 10000

/* 1-based Fortran column-major access for U(k1,k2,k3) with dims (2,101,5) */
#define U(arr,k1,k2,k3) arr[((k3)-1)*202 + ((k2)-1)*2 + (k1)]

int main(void) {
    /* Max flat index: (5-1)*202 + (101-1)*2 + 2 = 1010, need [1011] */
    double u1[1011], u2[1011], u3[1011];
    double du1[102], du2[102], du3[102];
    double spacer[40]; signel(spacer+1, 39);
    double a11 = spacer[1], a12 = spacer[2], a13 = spacer[3];
    double a21 = spacer[4], a22 = spacer[5], a23 = spacer[6];
    double a31 = spacer[7], a32 = spacer[8], a33 = spacer[9];
    double sig = spacer[34];

    /* Initialise: fill 1010 elements starting at index 1 */
    signel(&u1[1], 1010);
    signel(&u2[1], 1010);
    signel(&u3[1], 1010);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long l = 0; l < NREPS; l++) {
        long nl1 = 1, nl2 = 2;
        for (long kx = 2; kx <= 3; kx++) {
            for (long ky = 2; ky <= N - 1; ky++) {
                du1[ky] = U(u1,nl1,ky+1,kx) - U(u1,nl1,ky-1,kx);
                du2[ky] = U(u2,nl1,ky+1,kx) - U(u2,nl1,ky-1,kx);
                du3[ky] = U(u3,nl1,ky+1,kx) - U(u3,nl1,ky-1,kx);
                U(u1,nl2,ky,kx) = U(u1,nl1,ky,kx)
                    + a11*du1[ky] + a12*du2[ky] + a13*du3[ky]
                    + sig*(U(u1,nl1,ky,kx+1)
                    - 2.0*U(u1,nl1,ky,kx) + U(u1,nl1,ky,kx-1));
                U(u2,nl2,ky,kx) = U(u2,nl1,ky,kx)
                    + a21*du1[ky] + a22*du2[ky] + a23*du3[ky]
                    + sig*(U(u2,nl1,ky,kx+1)
                    - 2.0*U(u2,nl1,ky,kx) + U(u2,nl1,ky,kx-1));
                U(u3,nl2,ky,kx) = U(u3,nl1,ky,kx)
                    + a31*du1[ky] + a32*du2[ky] + a33*du3[ky]
                    + sig*(U(u3,nl1,ky,kx+1)
                    - 2.0*U(u3,nl1,ky,kx) + U(u3,nl1,ky,kx-1));
            }
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("u1(2,51,2) = %f\n", U(u1,2,51,2));
    return 0;
}

File: benchmarks/livermore/k08_adi.f
C     Kernel 8 -- ADI integration
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C
      PROGRAM K08
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION U1(2,101,5), U2(2,101,5), U3(2,101,5)
      DIMENSION DU1(101), DU2(101), DU3(101)
      DIMENSION SPACER(39)
C
C     Initialize spacer for constants
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      SCALED = 0.1D0
      DO 5 K = 1, 39
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          SPACER(K) = (BUZZ - FIZZ) * SCALED
    5 CONTINUE
      A11 = SPACER(1)
      A12 = SPACER(2)
      A13 = SPACER(3)
      A21 = SPACER(4)
      A22 = SPACER(5)
      A23 = SPACER(6)
      A31 = SPACER(7)
      A32 = SPACER(8)
      A33 = SPACER(9)
      SIG = SPACER(34)
C
C     Initialize U1, U2, U3 using SIGNEL pattern
C     C array: u1[2][101][5] = 1010 elements, stored row-major
C     Fortran array: U1(2,101,5) = 1010 elements, stored column-major
C     We fill in storage order (column-major for Fortran)
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 12 K3 = 1, 5
          DO 11 K2 = 1, 101
              DO 10 K1 = 1, 2
                  BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
                  FUZZ = -FUZZ
                  U1(K1,K2,K3) = (BUZZ - FIZZ) * SCALED
   10         CONTINUE
   11     CONTINUE
   12 CONTINUE
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 22 K3 = 1, 5
          DO 21 K2 = 1, 101
              DO 20 K1 = 1, 2
                  BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
                  FUZZ = -FUZZ
                  U2(K1,K2,K3) = (BUZZ - FIZZ) * SCALED
   20         CONTINUE
   21     CONTINUE
   22 CONTINUE
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 32 K3 = 1, 5
          DO 31 K2 = 1, 101
              DO 30 K1 = 1, 2
                  BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
                  FUZZ = -FUZZ
                  U3(K1,K2,K3) = (BUZZ - FIZZ) * SCALED
   30         CONTINUE
   31     CONTINUE
   32 CONTINUE
C
C     Kernel loop
C     C: nl1=0,nl2=1 (0-based) -> Fortran: NL1=1,NL2=2 (1-based)
C     C: kx=1..2 (0-based) -> Fortran: KX=2..3 (1-based)
C     C: ky=1..n-2 (0-based) -> Fortran: KY=2..N-1 (1-based)
C
      N = 100
      DO 200 REP = 1, 10000
          NL1 = 1
          NL2 = 2
          DO 110 KX = 2, 3
              DO 100 KY = 2, N - 1
                  DU1(KY) = U1(NL1,KY+1,KX) - U1(NL1,KY-1,KX)
                  DU2(KY) = U2(NL1,KY+1,KX) - U2(NL1,KY-1,KX)
                  DU3(KY) = U3(NL1,KY+1,KX) - U3(NL1,KY-1,KX)
                  U1(NL2,KY,KX) = U1(NL1,KY,KX)
     &                + A11*DU1(KY) + A12*DU2(KY) + A13*DU3(KY)
     &                + SIG*(U1(NL1,KY,KX+1)
     &                - 2.0D0*U1(NL1,KY,KX) + U1(NL1,KY,KX-1))
                  U2(NL2,KY,KX) = U2(NL1,KY,KX)
     &                + A21*DU1(KY) + A22*DU2(KY) + A23*DU3(KY)
     &                + SIG*(U2(NL1,KY,KX+1)
     &                - 2.0D0*U2(NL1,KY,KX) + U2(NL1,KY,KX-1))
                  U3(NL2,KY,KX) = U3(NL1,KY,KX)
     &                + A31*DU1(KY) + A32*DU2(KY) + A33*DU3(KY)
     &                + SIG*(U3(NL1,KY,KX+1)
     &                - 2.0D0*U3(NL1,KY,KX) + U3(NL1,KY,KX-1))
  100         CONTINUE
  110     CONTINUE
  200 CONTINUE
C
      WRITE(*,900) U1(2,51,2)
  900 FORMAT('K8 ADI: u1(2,51,2) = ', E25.15)
      END
